compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/MoveCompanionObjectFieldsLowering.kt

```diff
@@ -6,221 +6,132 @@
-            isStatic = true,
-            isFakeOverride = false
-        ).apply {
-            descriptor.bind(this)
-            parent = fieldParent
-            annotations.addAll(oldField.annotations)
-            metadata = oldField.metadata
-        }
-        val oldInitializer = oldField.initializer
-        if (oldInitializer != null) {
-            field.initializer = oldInitializer
-                .replaceThisByStaticReference(context, propertyParent, propertyParent.thisReceiver!!)
-                .deepCopyWithSymbols(field) as IrExpressionBody
+    private fun copyConstField(oldField: IrField, newParent: IrClass) =
+        newParent.addField {
+            updateFrom(oldField)
+            name = oldField.name
+            isStatic = true
+        }.apply {
+            parent = newParent
+            annotations += oldField.annotations
+            initializer = oldField.initializer?.deepCopyWithSymbols(this)
         }
-
-        return field
-    }
 }
 
-private fun IrElement.replaceFieldReferences(replacementMap: Map<IrFieldSymbol, IrFieldSymbol>) {
-    transformChildrenVoid(FieldReplacer(replacementMap))
-}
+private class RemapObjectFieldAccesses(val context: JvmBackendContext) : FileLoweringPass, IrElementTransformerVoid() {
+    override fun lower(irFile: IrFile) = irFile.transformChildrenVoid()
+
+    private fun IrField.remap(): IrField? =
+        correspondingPropertySymbol?.owner?.let(context.declarationFactory::getStaticBackingField)
 
-private class FieldReplacer(val replacementMap: Map<IrFieldSymbol, IrFieldSymbol>) : IrElementTransformerVoid() {
     override fun visitGetField(expression: IrGetField): IrExpression =
-        replacementMap[expression.symbol]?.let { newSymbol ->
-            IrGetFieldImpl(
-                expression.startOffset, expression.endOffset,
-                newSymbol,
-                expression.type,
-                /* receiver = */ null,
-                expression.origin,
-                expression.superQualifierSymbol
-            )
+        expression.symbol.owner.remap()?.let {
+            with(expression) {
+                IrGetFieldImpl(startOffset, endOffset, it.symbol, type, /* receiver = */ null, origin, superQualifierSymbol)
+            }
         } ?: super.visitGetField(expression)
 
     override fun visitSetField(expression: IrSetField): IrExpression =
-        replacementMap[expression.symbol]?.let { _ ->
-            IrSetFieldImpl(
-                expression.startOffset, expression.endOffset,
-                replacementMap.getValue(expression.symbol),
-                /* receiver = */ null,
-                visitExpression(expression.value),
-                expression.type,
-                expression.origin,
-                expression.superQualifierSymbol
-            )
+        expression.symbol.owner.remap()?.let {
+            with(expression) {
+                val newValue = value.transform(this@RemapObjectFieldAccesses, null)
```

Comment:
@udalov there was also another problem here due to the use of `visitExpression(expression.value)`, which only transforms the *children* of `expression.value`. This breaks for an obvious reason if `value` is `IrGetField` of a field made static, but to get `PropertiesToFieldsLowering` to generate such a read, the target of it needs to be in a yet-unlowered file, like so:

```kotline
// FILE: A.kt
object A {
    @JvmField
    val result: String

    init { result = Z.result } // -> IrSetField(A.result, IrGetField(Z.result))
}

// FILE: z.kt
object Z {
    @JvmField
    val result: String = "OK"
}
```